pkg/kubelet/cm/deviceplugin/pod_devices.go

```diff
@@ -17,241 +17,507 @@ limitations under the License.
+	ret := make(map[string]sets.String)
+	for _, cmap := range a.registry {
+		for _, rmap := range cmap {
+			for r, dinfo := range rmap {
+				if _, ok := ret[r]; !ok {
+					ret[r] = sets.NewString()
+				}
+				ret[r] = ret[r].Union(dinfo.ids)
+			}
+		}
 	}
 	return ret
 }
 
-func (pdev podDevices) insert(podUID, contName, resource string, devices sets.String, resp *pluginapi.AllocateResponse) {
-	if _, podExists := pdev[podUID]; !podExists {
-		pdev[podUID] = make(containerDevices)
+// purge updates the pod allocation registry by list of active pods.
+// The function returns a map from resource name to set of device IDs.
+func (a *devAllocationImpl) purge(activePods sets.String) {
+	a.Lock()
+	defer a.Unlock()
+
+	for p := range a.registry {
+		if !activePods.Has(p) {
+			delete(a.registry, p)
+		}
 	}
-	if _, contExists := pdev[podUID][contName]; !contExists {
-		pdev[podUID][contName] = make(resourceAllocateInfo)
+}
+
+// reserve processes the device requests from a Pod and return the set of
+// device IDs to allocate.
+func (a *devAllocationImpl) reserve(allDevices map[string]sets.String, req *allocRequest) (map[string]sets.String, error) {
+	a.Lock()
+	defer a.Unlock()
+
+	maxCount := make(map[string]int)
+	for container, rmap := range req.inits {
+		for resource, count := range rmap {
+			// filter out devices already allocated
+			allocated := a.devicesByContainer(req.pod, container, resource)
+			if allocated != nil {
+				glog.V(6).Infof("existing allocation found for container %q/%q: %s=%v",
+					req.pod, container, resource, allocated.List())
+				// TODO: allow reallocation when there is a need
+				if allocated.Len() != count {
+					return nil, fmt.Errorf("container %v/%v resource request changed: %v = %v to %v",
+						req.pod, container, resource, allocated.Len(), count)
+				}
+				req.inits[container][resource] = 0
+				continue
+			}
+			if m, ok := maxCount[resource]; ok && count <= m {
+				continue
+			}
+			maxCount[resource] = count
+		}
 	}
-	pdev[podUID][contName][resource] = deviceAllocateInfo{
-		deviceIds: devices,
-		allocResp: resp,
+
+	sumCount := make(map[string]int)
+	for container, rmap := range req.containers {
+		for resource, count := range rmap {
+			// filter out devices already allocated
+			allocated := a.devicesByContainer(req.pod, container, resource)
+			if allocated != nil {
+				glog.V(6).Infof("existing allocation found for container %q/%q: %s=%v",
+					req.pod, container, resource, allocated.List())
+				// TODO: allow reallocation when there is a need
+				if allocated.Len() != count {
+					return nil, fmt.Errorf("container %v/%v resource request changed: %v = %v to %v",
+						req.pod, container, resource, allocated.Len(), count)
+				}
+				continue
+			}
+			if s, found := sumCount[resource]; found {
+				count = count + s
+			}
+			sumCount[resource] = count
+
+			if m, ok := maxCount[resource]; ok && count <= m {
+				continue
+			}
+			maxCount[resource] = count
```

Comment:
Is the part in process? seems you plan to support device reuse between regular containers here, which is not by design;
Besides, I didn't see `sumCount` used somewhere.

Reply:
sumCount is used to add up all requests to a particular resource across all regular containers. It is only used during the iteration of regular containers. After this iteration, we get an accurate map `maxCount: resource=>count`. That map is all the (extended) resources requested by this Pod.
Device reuse across regular containers is not in the plan.

Reply:
Is what you said above already an implementation? I could only see the statement and accumulation of sumCount, but the result does not get used anywhere.
And then still calculate the device to allocate with maxCount, which indeed is an implementation of reuse. I didn't see any difference between init containers and regular containers in the process.

Reply:
sorry, I see what you mean

Reply:
:), reading this code needs a little bit mind shift from the existing retailing solution to the wholesale solution.